### PR TITLE
Break out scorer

### PR DESCRIPTION
**Pull request by henryre**
On the path to document-level scorers and optional pre-scoring operations (filtering, etc.)

### CODE REVIEW

File: snorkel/learning/utils.py [hunk 1]
```diff
@@ -1,14 +1,86 @@
...
 warnings.filterwarnings("ignore", module="matplotlib")
 
 
+class Scorer(object):
+    """Abstract type for scorers"""
+    def __init__(self, test_candidates, test_labels, gold_candidate_set):
+        """
+        test_candidates: candidates corresponding to test_labels
+        test_labels: ground truth labels for test candidates
+        gold_candidate_set: full set of gold labeled candidates
```
Comment (ajratner) [inline snorkel/learning/utils.py]: I think this has been a point of confusion for people, so let's specify the exact types here?  So:

- List of `Candidate`s
- `csrLabelMatrix`
- `CandidateSet`

Right?

File: snorkel/learning/utils.py [hunk 2]
```diff
@@ -1,14 +1,86 @@
...
+                    if test_label == 1:
+                        tp.add(candidate)
+                    else:
+                        fp.add(candidate)
+                else:
+                    if test_label == -1:
+                        tn.add(candidate)
+                    else:
+                        fn.add(candidate)
+            except KeyError:
```
Comment (ajratner) [inline snorkel/learning/utils.py]: I'm not sure we should catch this error... seems like inviting disaster in a scoring fn?

File: snorkel/learning/utils.py [hunk 3]
```diff
@@ -1,14 +1,86 @@
...
+        fn = set()
+
+        for i, candidate in enumerate(self.test_candidates):
+            try:
+                test_label_index = self.test_labels.get_row_index(candidate)
+                test_label   = self.test_labels[test_label_index, 0]
+
+                # Set unlabeled examples to -1 by default
+                if test_label == 0 and set_unlabeled_as_neg:
+                    test_label = -1
```
Comment (ajratner) [inline snorkel/learning/utils.py]: Looking at the code below, this `set_unlabeled_as_neg` flag isn't actually respected- it's always True.

I think we should actually just check to make sure all the test labels are in {-1,1}, else throw an error?

File: snorkel/learning/utils.py [hunk 4]
```diff
@@ -1,14 +1,86 @@
...
+            except KeyError:
+                test_label_array.append(-1)
+                if test_marginals[i] > b:
+                    fp.add(candidate)
+                else:
+                    tn.add(candidate)
+
+        # Print diagnostics chart and return error analysis candidate sets
+        if display:
+            predict = marginals_to_labels(test_marginals, b)
```
Comment (ajratner) [inline snorkel/learning/utils.py]: If we're going to use this `marginals_to_labels` fn, we should use consistently i.e. above too

File: snorkel/learning/utils.py [hunk 5]
```diff
@@ -1,14 +1,86 @@
...
+        gold_candidate_set: full set of gold labeled candidates
+        """
+        self.test_candidates = test_candidates
+        self.test_labels     = test_labels
+        self.gold_cs         = gold_candidate_set
+
+    def score(self, test_marginals, train_marginals=None, b=0.5, set_unlabeled_as_neg=True, display=True):
+        raise NotImplementedError()
+
+
```
Comment (ajratner) [inline snorkel/learning/utils.py]: I think we should also add the equivalent `RelationScorer` or whatever you had in mind, with `NotImplementedError()`, just so the intended design doesn't get forgotten if we forget to finish this?

File: snorkel/learning/utils.py (hunk 1 above)
Comment (henryre) [inline snorkel/learning/utils.py]: I'll respond verbally in a few seconds.

**Comment by henryre** [comment]
Agree with all of these points. `MentionScorer` was just a copy-paste from the old `score` method. We're going to redo scoring soon, so might be better to just pull this in now with the appropriate bug fixes

File: snorkel/learning/utils.py (hunk 1 above)
Comment (henryre) [inline snorkel/learning/utils.py]: But yes.

**Comment by ajratner** [comment]
Working on Emily's stuff / basic testing, merging in

**Pull request merged**